src/mlpack/methods/gmm/diagonal_em_fit_impl.hpp

```diff
@@ -0,0 +1,387 @@
+
+  // Now calculate the means, covariances, and weights.
+  // Initialize the means, covariances, and weights.
+  weights.zeros();
+  for (size_t i = 0; i < dists.size(); i++)
+  {
+    means[i].zeros(dists[i].Mean().n_elem);
+    covariances[i].zeros(dists[i].Covariance().n_elem);
+  }
+
+  // From the assignments, generate means, covariances, and weights.
+  for (size_t i = 0; i < observations.n_cols; i++)
+  {
+    const size_t cluster = assignments[i];
+
+    // Add this to the relevant mean.
+    means[cluster] += observations.col(i);
+
+    // Add this to the relevant covariance.
+    covariances[cluster] += observations.col(i) % observations.col(i);
+
+    // Add one to the weights to normalize parameters later.
+    weights[cluster]++;
+  }
+
+  // Normalize the mean.
+  for (size_t i = 0; i < dists.size(); i++)
+  {
+    means[i] /= (weights[i] > 1) ? weights[i] : 1;
+  }
+
+  // Calculate the covariances.
+  for (size_t i = 0; i < observations.n_cols; i++)
+  {
+    const size_t cluster = assignments[i];
+    const arma::vec diffs = observations.col(i) - means[cluster];
+    covariances[cluster] += diffs % diffs;
+  }
+
+  // Normalize the covariances.
+  for (size_t i = 0; i < dists.size(); i++)
+  {
+    covariances[i] /= (weights[i] > 1) ? weights[i] : 1;
+
+    std::swap(dists[i].Mean(), means[i]);
+
+    covariances[i] = arma::clamp(covariances[i], 1e-10, DBL_MAX);
+    dists[i].Covariance(std::move(covariances[i]));
+  }
+
+  // Normalize weights.
+  weights /= arma::accu(weights);
+}
+
+template<typename InitialClusteringType, typename CovarianceConstraintPolicy>
+double DiagonalEMFit<InitialClusteringType, CovarianceConstraintPolicy>::
+LogLikelihood(
+    const arma::mat& observations,
+    const std::vector<distribution::DiagonalGaussianDistribution>& dists,
+    const arma::vec& weights) const
+{
+  double logLikelihood = 0;
+
+  arma::vec phis;
+  arma::mat likelihoods(dists.size(), observations.n_cols);
+
+  for (size_t i = 0; i < dists.size(); ++i)
+  {
+    dists[i].Probability(observations, phis);
+    likelihoods.row(i) = weights(i) * trans(phis);
+  }
+  // Now sum over every point.
+  for (size_t j = 0; j < observations.n_cols; ++j)
+  {
+    if (accu(likelihoods.col(j)) == 0)
+      Log::Info << "Likelihood of point " << j << " is 0!  It is probably an "
+          << "outlier." << std::endl;
+    logLikelihood += log(accu(likelihoods.col(j)));
+  }
+
+  return logLikelihood;
+}
+
+template<typename InitialClusteringType, typename CovarianceConstraintPolicy>
+template<typename Archive>
+void DiagonalEMFit<InitialClusteringType, CovarianceConstraintPolicy>::
+serialize(Archive& ar, const unsigned int /* version */)
+{
+  ar & BOOST_SERIALIZATION_NVP(maxIterations);
+  ar & BOOST_SERIALIZATION_NVP(tolerance);
+  ar & BOOST_SERIALIZATION_NVP(clusterer);
+  ar & BOOST_SERIALIZATION_NVP(constraint);
+}
+
+// Armadillo uses uword internally as an OpenMP index type, which crashes
+// Visual Studio.
+#ifndef _WIN32
```

Comment:
Hmm, a bigger issue here is whether `ARMA_USE_OPENMP` is enabled on Windows.  If this does still crash Visual Studio, then we could modify `CMakeLists.txt` or `prereqs.hpp` or `core.hpp` to make sure that `ARMA_USE_OPENMP` is not defined when we are on Windows using Visual Studio.

If `ARMA_USE_OPENMP` is not defined, Windows should have no problems, meaning that we could always use the Armadillo implementation.  Unfortunately that would mean there's no reason to keep the handcrafted implementation you put in `DiagonalEMFit`... unless you can show it's faster than Armadillo's, in which case we should always use it. :)

Reply:
Ah, really good idea. :) I'll check it now! :)

Reply:
Hi @rcurtin, there is an issue, even though we don't define `ARMA_USE_OPENMP`. The ['gmm_diag'](https://gitlab.com/conradsnicta/armadillo-code/blob/9.300.x/include/armadillo_bits/gmm_diag_meat.hpp) class of Armadillo uses `uword` as an index type in `for` loop, regardless of using OpenMP. Visual Studio doesn't allow 'unsigned type index' in `for` loop (error: [C3016](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-2/compiler-error-c3016?view=vs-2017)). So I'm not sure we can always depend on Armadillo's `gmm_diag`.

I'll be merging the `EMFit` and `DiagonalEMFit`, please let me know what you think.